src/event_notifier.rs

```diff
@@ -0,0 +1,291 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//! Structure and wrapper functions for working with
+//! [`eventfd`](http://man7.org/linux/man-pages/man2/eventfd.2.html).
+//! [`pipe`](https://man7.org/linux/man-pages/man2/pipe.2.html).
+
+use std::mem;
+#[cfg(target_os = "macos")]
+use std::os::unix::io::RawFd;
+use std::os::unix::io::{AsRawFd, FromRawFd};
+use std::{fs::File, io, result};
+
+use libc::{c_void, dup, read, write};
+#[cfg(any(target_os = "linux", target_os = "android"))]
+pub use libc::{EFD_CLOEXEC, EFD_NONBLOCK, EFD_SEMAPHORE};
+#[cfg(target_os = "macos")]
+pub use libc::{O_CLOEXEC, O_NONBLOCK};
+
+/// A safe wrapper around Linux and MacOS
+/// [`eventfd`](http://man7.org/linux/man-pages/man2/eventfd.2.html).
+/// [`pipe`](https://man7.org/linux/man-pages/man2/pipe.2.html).
+#[derive(Debug)]
+pub struct EventNotifier {
+    rfd: File,
+    wfd: File,
+}
+
+impl EventNotifier {
+    /// Create a new EventNotifier with an initial value.
+    ///
+    /// # Arguments
+    ///
+    /// * `flag`: The initial value used for creating the `EventFd` or `PipeFd`.
+    ///   Refer to Linux [`eventfd`](http://man7.org/linux/man-pages/man2/eventfd.2.html) or
+    ///   [`pipe`](https://man7.org/linux/man-pages/man2/pipe.2.html).
+    /// # Examples
+    ///
+    /// ```
+    /// extern crate vmm_sys_util;
+    /// use vmm_sys_util::event_notifier::EventNotifier;
+    /// #[cfg(any(target_os = "linux", target_os = "android"))]
+    /// use vmm_sys_util::event_notifier::EFD_NONBLOCK;
+    /// #[cfg(target_os = "macos")]
+    /// use vmm_sys_util::event_notifier::O_NONBLOCK;
+    ///
+    /// #[cfg(any(target_os = "linux", target_os = "android"))]
+    /// EventNotifier::new(EFD_NONBLOCK).unwrap();
+    /// #[cfg(target_os = "macos")]
+    /// EventNotifier::new(O_NONBLOCK).unwrap();
+    /// ```
+    pub fn new(flag: i32) -> result::Result<EventNotifier, io::Error> {
+        #[cfg(any(target_os = "linux", target_os = "android"))]
+        {
+            // SAFETY: This is safe because eventfd merely allocated an eventfd for
+            // our process and we handle the error case.
+            let ret = unsafe { libc::eventfd(0, flag) };
+            if ret < 0 {
+                Err(io::Error::last_os_error())
+            } else {
+                Ok(EventNotifier {
+                    // SAFETY: This is safe because we checked ret for success and know
+                    // the kernel gave us an fd that we own.
+                    rfd: unsafe { File::from_raw_fd(ret) },
+                    wfd: unsafe { File::from_raw_fd(ret) },
```

Comment:
This won't work, as the `Drop` impls for both `File`s will try to close the FD `ret`, but the `File` drop impl aborts with "IO safety violation" if it encounters an already closed FD, as Rust-the-language declares that there should be a 1:1 correspondence between FDs and `File` objects.